metron-analytics/metron-profiler/src/main/java/org/apache/metron/profiler/bolt/ProfileSplitterBolt.java

```diff
@@ -0,0 +1,161 @@
+
+import java.io.UnsupportedEncodingException;
+import java.util.Map;
+
+import static java.lang.String.format;
+
+/**
+ * The bolt responsible for filtering incoming messages and directing
+ * each to the one or more bolts responsible for building a Profile.  Each
+ * message may be needed by 0, 1 or even many Profiles.
+ */
+public class ProfileSplitterBolt extends ConfiguredProfilerBolt {
+
+  protected static final Logger LOG = LoggerFactory.getLogger(ProfileSplitterBolt.class);
+
+  private OutputCollector collector;
+
+  /**
+   * JSON parser.
+   */
+  private transient JSONParser parser;
+
+  /**
+   * Executes Stellar code.
+   */
+  private StellarExecutor executor;
+
+  /**
+   * @param zookeeperUrl The Zookeeper URL that contains the configuration for this bolt.
+   */
+  public ProfileSplitterBolt(String zookeeperUrl) {
+    super(zookeeperUrl);
+  }
+
+  @Override
+  public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+    super.prepare(stormConf, context, collector);
+    this.collector = collector;
+    this.parser = new JSONParser();
+  }
+
+  @Override
+  public void execute(Tuple input) {
+    try {
+      doExecute(input);
+
+    } catch (IllegalArgumentException | ParseException | UnsupportedEncodingException e) {
+      LOG.error(format("Unexpected failure: message='%s', tuple='%s'", e.getMessage(), input), e);
+      collector.reportError(e);
+
+    } finally {
+      collector.ack(input);
+    }
+  }
+
+  private void doExecute(Tuple input) throws ParseException, UnsupportedEncodingException {
+
+    // retrieve the input message
+    byte[] data = input.getBinary(0);
+    JSONObject message = (JSONObject) parser.parse(new String(data, "UTF8"));
+
+    // ensure there is a valid profiler configuration
+    ProfilerConfig config = getProfilerConfig();
+    if(config == null) {
+      throw new IllegalArgumentException("Fatal: Unable to find valid profiler definition");
+    }
+
+    // apply the message to each of the profile definitions
+    for (ProfileConfig profile: config.getProfiles()) {
+      applyProfile(profile, input, message);
+    }
+  }
+
+  /**
+   * Applies a message to a Profile definition.
+   * @param profile The profile definition.
+   * @param input The input tuple that delivered the message.
+   * @param message The message that may be needed by the profile.
+   */
+  private void applyProfile(ProfileConfig profile, Tuple input, JSONObject message) throws ParseException, UnsupportedEncodingException {
+
+    // is this message needed by this profile?
+    String onlyIf = profile.getOnlyif();
+    if (executor.execute(onlyIf, message, Boolean.class)) {
+
+      // what is the name of the entity in this message?
+      String entity = executor.execute(profile.getForeach(), message, String.class);
+
+      // emit a message for the bolt responsible for building this profile
+      collector.emit(input, new Values(entity, profile, copyMessage(input)));
```

Comment:
No.  I don't think I need a deep copy there.  I started with that early on in the build, but it is not needed now.